site/content/blog/2020-06-10-opentracing.md

```diff
@@ -0,0 +1,576 @@
+    if (typeof target[prop] === "function") {
+      return function(...args) {
+        console.log(`'${prop} 'called with arguments: `, ...arguments);
+        return target[prop](...arguments);
+      };
+    }
+  }
+};
+
+const timerHandler = {
+  get: function(target, prop, receiver) {
+    if (typeof target[prop] === "function") {
+      return function(...args) {
+        console.log(`starting '${prop}'`);
+        const t1 = window.performance.now();
+
+        const res = target[prop](...arguments);
+        const t2 = window.performance.now();
+
+        console.log(`'${prop}' took ${t2 - t1}ns`);
+        return res;
+      };
+    }
+  }
+};
+
+const proxy = new Proxy(cow, tracerHandler);
+const proxy2 = new Proxy(proxy, timerHandler);
+console.log(proxy2.feed(20));
+console.log(proxy2.speak(3));
+```
+
+Unfortunately, in Go, there's no way to do this in a performant manner (the regular approach would involve using reflection and that can have a serious performance impact on the underlying code)
+
+## Our solution
+
+Implementation of decorator pattern that we chose involved 3 parts:
+
+1. Struct embedding
+2. Code parsing using AST
+3. Code generation using templates
+
+### Struct embedding
+
+Since Go is not object-oriented, it uses Struct embedding as a type of inheritance. Here's a small example:
+
+```go
+type Animal struct{
+	Name string
+}
+
+type Cow struct{
+	Animal
+}
+
+func (c Cow) Speak() {
+	fmt.Printf("Moo, I am a %s", c.Animal.Name)
+}
+
+func main() {
+	a := Animal{Name:"Cow"}
+	c := Cow{Animal:a}
+	c.Speak()
+}
+```
+
+How does struct embedding help us in implementation of decorator pattern?
+
+```go
+
+type IAnimal interface {
+	Speak(x int)
+}
+
+type Animal struct {
+	Name string
+}
+
+type TraceAnimal struct {
+	IAnimal
+}
+
+type MeasureAnimal struct {
+	IAnimal
+}
+
+func (c Animal ) Speak(x int) {
+	fmt.Println(strings.Repeat("I am a " + c.Name + " ",x))
+}
+
+func (c TraceAnimal) Speak(x int) {
+	fmt.Printf("Running Speak(x) function with x=%d!\n",x)
+	c.IAnimal.Speak(x)
+}
+
+func (c MeasureAnimal) Speak(x int) {
+	fmt.Println("Timing Speak() function...")
+	t := time.Now()
+	c.IAnimal.Speak(x)
+	
+	fmt.Printf("Speak(%d) took %s\n", x, time.Since(t))
+}
+
+func main() {
+	a := Animal{Name: "Cow"}
+	c := TraceAnimal {IAnimal: a}
+	d := MeasureAnimal{IAnimal: c}
+	d.Speak(2)
+}
+```
+
+Running the following code will yield:
+
+```
+Timing Speak() function...
+Running Speak(x) function with x=2!
+I am a Cow I am a Cow 
+Speak(2) took 0s
+```
+
+So we've basically implemented two decorators over original `Speak()` method, first we started timing the execution in `MeasureAnimal`, then passed it too `TraceAnimal` which in turn called the actual `Speak()` implementation.
+
+This works great and stays performant since we don't use any dynamic techniques here like `reflection`, however this is very verbose and requires us to write a lot of wrapper code, and that's no fun at all. We can do better!
```

Comment:
reflection need not be in backticks

Reply:
Should it be capitalized or italicized?

Reply:
It should be a regular word. Readers would know it.